README.md
![](logo.png)

[![Build Status](https://travis-ci.org/britzl/defold-googleanalytics.svg?branch=master)](https://travis-ci.org/britzl/defold-googleanalytics)

# Google Analytics for Defold
This is a Lua implementation of [Google Analytics](https://www.google.com/analytics) for the [Defold game engine](http://www.defold.com). The project is provided as a Defold library project for easy integration into Defold games. The implementation is loosely based on the design of the Google Analytics Android SDK, but with several simplifications thanks to the dynamic and flexible nature of Lua.

This Lua implementation uses the [Google Analytics Measurement Protocol](https://developers.google.com/analytics/devguides/collection/protocol/v1/) to make direct calls to the Google Analytics servers. On top of these raw calls the implementation also adds support for offline tracking, automatic crash/exception reporting and automatic retrieval of relevant tracking parameters such as app name, app id, language, screen resolution and so on.

## Installation
You can use Google Analytics in your own project by adding this project as a [Defold library dependency](http://www.defold.com/manuals/libraries/). Open your game.project file and in the dependencies field under project add:

	https://github.com/britzl/defold-googleanalytics/archive/master.zip

Or point to the ZIP file of a [specific release](https://github.com/britzl/defold-googleanalytics/releases).

## Configuration
Before you can use Google Analytics in your project you need to add your analytics tracking ID to game.project. Open game.project as a text file and create a new section:

	[googleanalytics]
	tracking_id = UA-1234567-1

Additional optional values are:

	[googleanalytics]
	dispatch_period = 1800
	queue_save_period = 60
	verbose = 1

`dispatch_period` is the interval, in seconds, at which tracking data is sent to the server.

`queue_save_period` is the minimum interval, in seconds, at which tracking data is saved to disk.

`verbose` set to 1 will print some additional data about when and how many hits are sent to Google Analytics. Set to 0 or omit the value to not print anything.

## Usage
Once you have added your tracking ID to game.project you're all set to start sending tracking data:

	local ga = require "googleanalytics.ga"

	function init(self)
		ga.get_default_tracker().screenview("my_cool_screen")
	end

	function update(self, dt)
		ga.update()
	end

	function on_input(self, action_id, action)
		if gui.pick_node(node1, action.x, action.y) and action.pressed then
			ga.get_default_tracker().event("category", "action")
		end

		if gui.pick_node(node2, action.x, action.y) and action.pressed then
			local time = socket.gettime()
			http.request("http://d.defold.com/stable/info.json", "GET", function(self, id, response)
				local millis = math.floor((socket.gettime() - time) * 1000)
				ga.get_default_tracker().timing("http", "get", millis)
			end)
		end
	end

Note that all tracking arguments of type string will be automatically URL encoded by the library.

## Supported hit types
This implementation supports the following hit types:

* Event - `ga.get_default_tracker().event(category, action, label, value)`
* Screen View - `ga.get_default_tracker().screenview(screen_name)`
* Timing - `ga.get_default_tracker().timing(category, variable, time, label)`
* Exception - `ga.get_default_tracker().exception(description, is_fatal)`, also see section on automatic crash/exception tracking

You can also register a raw hit where you specify all parameters yourself:

	ga.get_default_tracker().raw("v=1&tid=UA-123456-1&cid=5555&t=pageview&dp=%2Fpage")

A set of base parameters such as screen dimensions, uuid, application name etc are provided in `base_params` on the tracker instance. These can be useful when creating the params for a raw hit:

	print(ga.get_default_tracker().base_params)

	v=1&ds=app&cid=b80e6164-fc1f-4d76-cdae-dfb7e9a9507c&tid=UA-87977671-1&vp=1280x720&ul=en&an=Google_Analytics&aid=Google_AnalyticsDarwin&av=0.9

## Automatic crash/exception tracking
You can let Google Analytics automatically send tracking data when your app crashes. The library can handle soft crashes (ie when your Lua code crashes) using [sys.set_error_handler](http://www.defold.com/ref/sys/#sys.set_error_handler:error_handler) and hard crashes (ie when the Defold engine crashes) using [crash API](http://www.defold.com/ref/crash/). Enable automatic crash tracking like this:

	local ga = require "googleanalytics.ga"

	function init(self)
		ga.get_default_tracker().enable_crash_reporting(true)
	end

## License
This library is released under the same [Terms and Conditions as Defold](http://www.defold.com/about-terms/).

## Third party tools and modules used
The library uses the following modules:

* [json.lua by rxi](https://github.com/rxi/json.lua) (MIT License)
* [uuid.lua by Tieske](https://github.com/Tieske/uuid) (Apache 2.0)
* [url_encode() from Lua String Recipes](http://lua-users.org/wiki/StringRecipes)

The example project uses:

* [Dirty Larry UI library](https://github.com/andsve/dirtylarry)
* [Spineboy animation from the Spine animation tool](https://github.com/EsotericSoftware/spine-superspineboy).


CHANGELOG.md
## Defold Google Analytics 1.3.2 [britzl released 2017-11-23]
FIX: Operator precedence fix for version tracking

## Defold Google Analytics 1.3.1 [britzl released 2017-08-24]
Use project version default value if none exists in game.project

## Defold Google Analytics 1.3 [britzl released 2017-01-24]
The library will no longer attempt to set a user-agent string on HTML5 builds since this is prevented by modern browsers (manifesting itself as a non-critical 'Refused to set unsafe header "User-Agent"' message).


## Defold Google Analytics 1.2 [britzl released 2016-12-20]
Added user agent request header when posting analytics data. This will fix the issue with Real-Time data in the Google Analytics Dashboard always showing all visitors as Desktop.


## Defold Google Analytics 1.1 [britzl released 2016-12-19]
Fixed issues when saving the queue on Android


## First public release [britzl released 2016-11-29]




tests/test_ga.lua
local mock = require "deftest.mock.mock"
local mock_fs = require "deftest.mock.fs"
local mock_time = require "deftest.mock.time"
local file = require "googleanalytics.internal.file"

return function()
	local ga
	local queue
	local tracker
	
	describe("ga", function()
		local mocked_sys_config_values = {}
		local function mock_sys_config_value(key, value)
			mocked_sys_config_values[key] = value
		end
	
		before(function()
			mock.mock(sys)
			mock_fs.mock()
			mock_time.mock()

			ga = require "googleanalytics.ga"
			tracker = require "googleanalytics.tracker"
			queue = require "googleanalytics.internal.queue"
			mock.mock(queue)
			mock.mock(tracker)
			--queue.dispatch.replace(function() end)
			sys.get_config.replace(function(key, default)
				local value = mocked_sys_config_values[key]
				if value and type(value) == "string" then
					return value
				elseif value ~= nil and type(value) ~= "string" then
					return default
				else
					return sys.get_config.original(key, default)
				end
			end)
		end)

		after(function()
			mock.unmock(sys)
			mock.unmock(queue)
			mock.unmock(tracker)
			mock_fs.unmock()
			mock_time.unmock()
			
			package.loaded["googleanalytics.internal.queue"] = nil
			package.loaded["googleanalytics.ga"] = nil
			package.loaded["googleanalytics.tracker"] = nil
		end)

		it("has a default tracker instance", function()
			local t = ga.get_default_tracker()
			assert(t)
			assert(ga.get_default_tracker() == t)
		end)

		it("should read tracking id from game.project", function()
			mock_sys_config_value("googleanalytics.tracking_id", "UA-123456-7")
		
			local t = ga.get_default_tracker()
			assert(t)
			assert(tracker.create.params[1] == "UA-123456-7")
		end)

		it("should read dispatch period from game.project", function()
			mock_sys_config_value("googleanalytics.dispatch_period", "12345")
		
			package.loaded["googleanalytics.ga"] = nil
			ga = require "googleanalytics.ga"
			assert(ga.dispatch_period == 12345)
		end)

		it("should use a default dispatch period if none is provided game.project", function()
			mock_sys_config_value("googleanalytics.dispatch_period", false)
		
			package.loaded["googleanalytics.ga"] = nil
			ga = require "googleanalytics.ga"
			assert(ga.dispatch_period == 30 * 60)
		end)

		it("should let the queue dispatch hits at regular intervals", function()
			mock_time.set(0)
			ga.update()
			assert(queue.dispatch.calls == 1)
			ga.update()
			assert(queue.dispatch.calls == 1)

			mock_time.elapse(ga.dispatch_period)
			
			ga.update()
			assert(queue.dispatch.calls == 2)
		end)

		it("should not automatically dispatch hits if set to manually dispatch them", function()
			ga.dispatch_period = 0
			ga.update()
			assert(queue.dispatch.calls == 0)
		end)
	end)
end

tests/test_file.lua
local mock_fs = require "deftest.mock.fs"
local file = require "googleanalytics.internal.file"

return function()
	describe("file", function()
		local file
	
		before(function()
			mock_fs.mock()
			file = require "googleanalytics.internal.file"
		end)

		after(function()
			mock_fs.unmock()
			package.loaded["googleanalytics.internal.file"] = nil
		end)

		it("should be able to provide a full path to a save file", function()
			assert(file.get_save_file_name("foobar"))
		end)

		it("should be able write to a file", function()
			local name1 = "foo"
			local name2 = "bar"
			file.save(name1, "some data")
			file.save(name2, "some other data")

			assert(mock_fs.has_file(file.get_save_file_name(name1)))
			assert(mock_fs.has_file(file.get_save_file_name(name2)))
			assert(mock_fs.get_file(file.get_save_file_name(name1)) == "some data")
			assert(mock_fs.get_file(file.get_save_file_name(name2)) == "some other data")
		end)

		it("should write to a temporary file and then move it if successful", function()
			file.save("foobar", "some data")

			assert(os.rename.calls == 1)
			assert(os.remove.calls == 1)
		end)

		it("should not write a partial file to disk", function()
			file.save("foobar", "some data")
			mock_fs.fail_writes(true)
			assert_error(function() file.save("foobar", "some other data") end)
			
			assert(mock_fs.get_file(file.get_save_file_name("foobar")) == "some data")
		end)

		it("should be able to load an existing file", function()
			local name1 = "foo"
			local name2 = "bar"
			file.save(name1, "some data")
			file.save(name2, "some other data")
			
			assert(file.load(name1) == "some data")
			assert(file.load(name2) == "some other data")
		end)

		it("should not crash when loading a file that does not exist", function()
			local data, err = file.load("foobar")
			assert(not data and err)
		end)
	end)
end

tests/test_uuid.lua
return function()
	local uuid_generator
	
	describe("uuid", function()
		before(function()
			uuid_generator = require "googleanalytics.internal.uuid"
		end)

		after(function()
			package.loaded["googleanalytics.internal.uuid"] = nil
		end)

		it("should generate uuids", function()
			uuid_generator.seed()
			local uuid = uuid_generator()
			assert(uuid)
			-- d902fd39-bbd3-41ad-c1e9--0ee49473f3e0
			assert(uuid:match("%x%x%x%x%x%x%x%x%-%x%x%x%x%-%x%x%x%x%-%x%x%x%x%-%x%x%x%x%x%x%x%x%x%x%x%x") == uuid)
		end)
	end)
end

googleanalytics/tracker.lua
local uuid_generator = require "googleanalytics.internal.uuid"
local queue = require "googleanalytics.internal.queue"
local file = require "googleanalytics.internal.file"

local M = {}

local UUID_FILENAME = "__ga_uuid"

local function url_encode(str)
	if (str) then
		str = string.gsub (str, "\n", "\r\n")
		str = string.gsub (str, "([^%w %-%_%.%~])", function (c) return string.format ("%%%02X", string.byte(c)) end)
		str = string.gsub (str, " ", "+")
	end
	return str	
end

local function url_decode(str)
	str = string.gsub (str, "+", " ")
	str = string.gsub (str, "%%(%x%x)", function(h) return string.char(tonumber(h,16)) end)
	str = string.gsub (str, "\r\n", "\n")
	return str
end


local function get_uuid()
	local uuid, err = file.load(UUID_FILENAME)
	if not uuid then
		uuid_generator.seed()
		uuid = uuid_generator()
		file.save(UUID_FILENAME, uuid)
	end
	return uuid
end

local function get_application_name()
	return sys.get_config("project.title"):gsub(" ", "_")
end

local function get_application_id()
	local APPLICATION_IDS = {
		["Android"] = sys.get_config("android.package"),
		["iPhone OS"] = sys.get_config("ios.bundle_identifier"),
		["Darwin"] = sys.get_config("osx.bundle_identifier"),
	}
	local system_name = sys.get_sys_info().system_name
	return APPLICATION_IDS[system_name] or (get_application_name() .. system_name)
end


--- Create a tracker instance
-- @param tracking_id Tracking id from the Google Analytics admin dashboard
-- @return Tracker instance
function M.create(tracking_id)
	local tracker = {
		-- https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters
		base_params = "v=1&ds=app"
		.. "&cid=" .. get_uuid()
		.. "&tid=" .. tracking_id
		.. "&vp=" .. sys.get_config("display.width") .. "x" .. sys.get_config("display.height")
		.. "&ul=" .. sys.get_sys_info().device_language
		.. "&an=" .. url_encode(get_application_name())
		.. "&aid=" .. url_encode(get_application_id())
		.. "&av=" .. (sys.get_config("project.version") or "1.0"),
	}
	
	local event_params = tracker.base_params .. "&t=event"
	local timing_params = tracker.base_params .. "&t=timing"
	local screenview_params = tracker.base_params .. "&t=screenview"
	local exception_params = tracker.base_params .. "&t=exception"
	
	--- Enable or disable crash reporting
	-- If enabled the tracker will automatically get Lua soft crashes using
	-- the sys.set_error_handler() function and track them as non-fatal crashes.
	-- It will also, when called, load any previous hard crash using
	-- crash.load_previous() and track that as a fatal crash.
	-- @param enabled Set to true to enable automatic crash reporting
	function tracker.enable_crash_reporting(enabled)
		if enabled then
			sys.set_error_handler(function(source, message, traceback)
				tracker.exception(message, false)
			end)
			local handle = crash.load_previous()
			if handle then
				tracker.exception(crash.get_extra_data(handle), true)
				crash.release(handle)
			end
		else
			sys.set_error_handler(function() end)
		end
	end
	
	--- Raw tracking
	-- @params params A valid Google Analytics parameter string
	function tracker.raw(params)
		assert(params and type(params) == "string", "You must provide some params (of type string)")
		queue.add(params)
	end
	
	--- Exception tracking
	-- @param description Specifies the description of an exception. Optional.
	-- @param is_fatal Specifies whether the exception was fatal. Optional.
	function tracker.exception(description, is_fatal)
		assert(not description or type(description) == "string", "Description must be nil or of type string")
		assert(is_fatal == nil or type(is_fatal) == "boolean", "Is_fatal must be nil or of type boolean")
		queue.add(exception_params .. (is_fatal ~= nil and ("&exf=" .. (is_fatal and "1" or "0")) or "") .. (description and ("&exd=" .. url_encode(description)) or ""))
	end
	
	--- Event tracking
	-- @param category Specifies the event category. Must not be empty.
	-- @param action Specifies the event action. Must not be empty.
	-- @param label Specifies the event label. Optional.
	-- @param value Specifies the event value. Optional. Non-negative.
	function tracker.event(category, action, label, value)
		assert(category and type(category) == "string" and #category > 0, "You must provide a category (of type string, must not be empty)")
		assert(action and type(action) == "string" and #action > 0, "You must provide an action (of type string, must not be empty)")
		assert(not label or type(label) == "string", "Label must be nil or of type string")
		assert(not value or (type(value) == "number" and value >= 0), "Value must be nil or a positive number")
		queue.add(event_params .. "&ec=" .. url_encode(category) .. "&ea=" .. url_encode(action) .. (label and ("&el=" .. url_encode(label)) or "") .. (value and ("&ev=" .. tostring(value)) or ""))
	end
	
	--- Screenview
	-- @param screen_name Specified the screen name of a screenview hit
	function tracker.screenview(screen_name)
		assert(screen_name and type(screen_name) == "string", "You must specify a screen name (of type string)")
		queue.add(screenview_params .. "&cd=" .. url_encode(screen_name))
	end
	
	--- User timing
	-- @param category Specifies the user timing category
	-- @param variable Specifies the user timing variable
	-- @param time Specifies the user timing value (milliseconds)
	-- @param label Specifies the user timing label. Optional.
	function tracker.timing(category, variable, time, label)
		assert(category and type(category) == "string", "You must provide a category (of type string)")
		assert(variable and type(variable) == "string", "You must provide a variable (of type string)")
		assert(time and type(time) == "number" and time >= 0, "You must provide a time (as a positive number)")
		assert(not label or type(label) == "string", "Label must be nil or a string")
		queue.add(timing_params .. "&utc=" .. url_encode(category) .. "&utv=" .. url_encode(variable) .. "&utt=" .. tostring(time) .. (label and ("&utl=" .. url_encode(label)) or ""))
	end

	return tracker
end


return M


googleanalytics/internal/file.lua
local M = {}


--- Get the full path to a save file
-- @param name Name of file to get full path for
-- @return The full path to the specified file
function M.get_save_file_name(name)
	local application_name = sys.get_config("project.title"):gsub(" ", "_")
	return sys.get_save_file(application_name, name)
end


--- Load a file from disk
-- @param name Name of file to load
-- @return contents File contents or nil if file could not be read
-- @return error_message Error message if file could not be read
function M.load(name)
	assert(name, "You must provide a file name")
	local filename = M.get_save_file_name(name)
	local file, err = io.open(filename, "r")
	if not file then
		return nil, err
	end
	local contents = file:read("*a")
	if not contents then
		return nil, "Unable to read file"
	end
	return contents
end


--- Save a file to disk
-- Saves are atomic and first written to a temporary file
-- If the file already exists it will be overwritten
-- @param name Name of file to write data to
-- @param data The data to write
-- @return success
-- @return error_message
function M.save(name, data)
	assert(name, "You must provide a file name")
	assert(data, "You must provide some data")
	assert(type(data) == "string", "You can only write strings")
	local tmpname = M.get_save_file_name("___gatmp")
	os.remove(tmpname)
	local file, err = io.open(tmpname, "w+")
	if not file then
		return nil, err
	end
	file:write(data)
	file:close()

	local filename = M.get_save_file_name(name)
	os.remove(filename)
	return os.rename(tmpname, filename)
end

return M

googleanalytics/internal/queue.lua
--- Queue of Google Analytics hit tracking
-- The queue provides functionality to add data to the queue and to send the
-- data to the Google Analytics servers.
-- The queue is persisted to disk. It is loaded from disk when this module
-- is initialised and written periodically.
-- The queue will hold a configurable amount of hit data. If the limit is
-- reached the oldest data will be thrown away.

local file = require "googleanalytics.internal.file"
local json_encode = require "googleanalytics.internal.json_encode"
local user_agent = require "googleanalytics.internal.user_agent"

local M = {
	last_dispatch_time = nil,
	last_save_time = nil,
	minimum_save_period = tonumber(sys.get_config("googleanalytics.queue_save_period", 5 * 60)),
}


-- payload rules, https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide
local MAX_HITS_PER_PAYLOAD = 20				-- A maximum of 20 hits can be specified per request.
local MAX_HIT_SIZE = 8 * 1024				-- No single hit payload can be greater than 8K bytes.
local MAX_TOTAL_PAYLOAD_SIZE = 16 * 1024	-- The total size of all hit payloads cannot be greater than 16K bytes.

local QUEUE_FILENAME = "__ga_queue"

local q = {}

local queue_data = file.load(QUEUE_FILENAME)
if queue_data then
	local decoded_queue, err = pcall(json.decode, json)
	if not err then
		q = decoded_queue
	end
end


local function sort()
	table.sort(q, function(a, b)
		return a.time < b.time
	end)
end


function M.log(...)
	-- no-op
end

if sys.get_config("googleanalytics.verbose") == "1" then
	M.log = function(s, ...)
		if select("#", ...) > 0 then
			print(s:format(...))
		else
			print(s)
		end
	end
end

--- Add tracking parameters to queue
-- @param params
function M.add(params)
	assert(params, "You must provide params")
	if #params <= MAX_HIT_SIZE then
		table.insert(q, { time = socket.gettime(), params = params })
	end
	
	if #q > 0 and (not M.last_save_time or (socket.gettime() >= (M.last_save_time + M.minimum_save_period))) then
		local ok, err = pcall(function()
			assert(file.save(QUEUE_FILENAME, json_encode.encode(q)))
		end)
		if not ok then
			M.log("ERR: Something went wrong while saving Google Analytics data", err)
			M.dispatch()
			q = {}
			return
		end
		M.last_save_time = socket.gettime()
	end
end

--- Dispatch all stored hits to Google Analytics
function M.dispatch()
	M.last_dispatch_time = socket.gettime()
	if #q == 0 then
		M.log("Nothing to send")
		return
	end
	
	local hits_to_retry = {}
	while #q > 0 do
		
		local hits_in_flight = {}
		local payload = {}
		local payload_size = 0
		for _=1,math.min(MAX_HITS_PER_PAYLOAD, #q) do
			local hit = q[1]
			local queue_time = math.floor((socket.gettime() - hit.time) * 1000)
			local qt = "&qt=" .. tostring(queue_time)
			local hit_size = #hit.params + #qt
			if payload_size + hit_size <= MAX_TOTAL_PAYLOAD_SIZE then
				local data = table.remove(q, 1)
				table.insert(hits_in_flight, data)
				payload[#payload + 1] = data.params .. qt
				payload_size = payload_size + hit_size
			end
		end
		
		if #payload > 0 then
			local post_data = table.concat(payload, "\n")
			local headers = { ["User-Agent"] = user_agent.get() }
			M.log("Sending %d hit(s) to Google Analytics", #hits_in_flight)
			http.request("https://www.google-analytics.com/batch", "POST", function(self, id, response)
				if response.status < 200 or response.status >= 300 then
					M.log("ERR: Problem when sending hits to Google Analytics. Code: ", response.status)
					for i=1,#hits_in_flight do
						hits_to_retry[#hits_to_retry + 1] = hits_in_flight[i]
					end
				end
			end, headers, post_data)
		end
	end

	-- re-add any hits that we failed to send due to http errors
	if #hits_to_retry > 0 then
		for i=1,#hits_to_retry do
			q[#q + 1] = hits_to_retry[i]
		end
		sort()
	end
end



return M

googleanalytics/internal/user_agent.lua
--- Utility module to create a reasonably well formatted user agent string based
-- on the platform dmengine is running on.

local M = {}

--- Get a user agent string
-- @return User agent string
function M.get()
	local sys_info = sys.get_sys_info()
	local device_model = sys_info.device_model
	local manufacturer = sys_info.manufacturer
	local system_name = sys_info.system_name
	local system_version = sys_info.system_version
	local system_version_underscore = system_version:gsub("%.", "_")
	local api_version = sys_info.api_version
	local device_language = sys_info.device_language

	local engine_info = sys.get_engine_info()
	local engine_version = engine_info.version

	local user_agent
	if sys_info.system_name == "Android" then
		user_agent = ("Mozilla/5.0 (Linux; Android %s; %s %s) AppleWebKit/537.36 (KHTML, like Gecko) Defold/%s"):format(system_version, manufacturer, device_model, engine_version)
	elseif sys_info.system_name == "iPhone OS" then
		if sys_info.device_model:match("iPad.*") == sys_info.device_model then
			user_agent = ("Mozilla/5.0 (iPad; CPU OS %s like Mac OS X) AppleWebKit/537.36 (KHTML, like Gecko) Defold/%s"):format(system_version_underscore, engine_version)
		elseif sys_info.device_model:match("iPod.*") == sys_info.device_model then
			user_agent = ("Mozilla/5.0 (iPod; CPU OS %s like Mac OS X) AppleWebKit/537.36 (KHTML, like Gecko) Defold/%s"):format(system_version_underscore, engine_version)
		else
			user_agent = ("Mozilla/5.0 (iPhone; CPU iPhone OS %s like Mac OS X) AppleWebKit/537.36 (KHTML, like Gecko) Defold/%s"):format(system_version_underscore, engine_version)
		end
	elseif sys_info.system_name == "Darwin" then
		user_agent = ("Mozilla/5.0 (Macintosh; Intel Mac OS X %s) AppleWebKit/602.3.12 (KHTML, like Gecko) Defold/%s"):format(system_version, engine_version)
	elseif sys_info.system_name == "Windows" then
		user_agent = ("Mozilla/5.0 (MSIE 10.0; Windows NT %s; Trident/5.0) Defold/%s"):format(system_version, engine_version)
	elseif sys_info.system_name == "Linux" then
		user_agent = ("Mozilla/5.0 (X11; Linux x86_64; Debian) AppleWebKit/537.36 (KHTML, like Gecko) Defold/%s"):format(engine_version)
	else
		user_agent = ("Mozilla/5.0 (%s; %s; %s %s) Defold/%s"):format(system_name, system_version, device_model, manufacturer, engine_version)
	end
	return user_agent
end


return M

googleanalytics/ga.lua
--- Main module for the Google Analytics implementation for the Defold game
-- engine.
--
-- @usage
-- local ga = require "googleanalytics.ga"
-- 
-- function init(self)
-- 	ga.get_default_tracker().screenview("my_cool_screen")
-- end
-- 
-- function update(self, dt)
-- 	ga.update()
-- end
-- 
-- function on_input(self, action_id, action)
-- 	if gui.pick_node(node, action.x, action.y) and action.pressed then
-- 		go.get_default_tracker().event("category", "action"))
-- 	end
-- end


local tracker = require "googleanalytics.tracker"
local queue = require "googleanalytics.internal.queue"

local M = {
	dispatch_period = tonumber(sys.get_config("googleanalytics.dispatch_period", 30 * 60)),
}

local default_tracker = nil

--- Get the default tracker
-- @return Tracker instance
function M.get_default_tracker()
	if not default_tracker then
		local tracking_id = sys.get_config("googleanalytics.tracking_id")
		assert(tracking_id, "You must set tracking_id in section [googleanalytics] in game.project before using this module")
		default_tracker = tracker.create(tracking_id)
	end
	return default_tracker
end

--- Dispatch hits to Google Analytics
function M.dispatch()
	queue.dispatch()
end

--- Update the Google Analytics module.
-- This will check if automatic dispatch of hits are enabled and if so, if it is
-- time to dispatch stored hits.
function M.update()
	-- manual dispatch only?
	if M.dispatch_period <= 0 then
		return
	end
	
	if not queue.last_dispatch_time or (socket.gettime() >= (queue.last_dispatch_time + M.dispatch_period)) then 
		M.dispatch()
	end
end


return M